File: 1/2020.10.27/6道股票问题/309 冷却期/python/solution.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        dp0, dp1 = 0, float('-inf')
        dp_prev0 = 0
        for price in prices:
            tmp = dp0
            dp0 = max(dp0, dp1 + price)
            dp1 = max(dp1, dp_prev0 - price)
            dp_prev0 = tmp
        return dp0

File: 1/2020.10.27/6道股票问题/309 冷却期/java/solution.java
class Solution {
    public int maxProfit(int[] prices) {
        int dp0 = 0, dp1 = Integer.MIN_VALUE;
        int dpPrev0 = 0;
        for (int price : prices) {
            int tmp = dp0;
            dp0 = Math.max(dp0, dp1 + price);
            dp1 = Math.max(dp1, dpPrev0 - price);
            dpPrev0 = tmp;
        }
        return dp0;
    }
}
